src/mongo/client/dbclient.cpp

```diff
@@ -1066,15 +1070,175 @@ namespace mongo {
+            md5_finish(&st, d);
+        }
+        return digestToString( d );
+    }
+
+    bool DBClientWithCommands::copyDatabase(const string &fromdb,
+                                            const string &todb,
+                                            const string &fromhost,
+                                            const string &mechanism,
+                                            const string &username,
+                                            const string &password,
+                                            BSONObj *info) {
         BSONObj o;
         if ( info == 0 ) info = &o;
-        BSONObjBuilder b;
-        b.append("copydb", 1);
-        b.append("fromhost", fromhost);
-        b.append("fromdb", fromdb);
-        b.append("todb", todb);
-        return runCommand("admin", b.done(), *info);
+        BSONObjBuilder copydbCmd;
+        copydbCmd.append("copydb", 1);
+        copydbCmd.append("fromhost", fromhost);
+        copydbCmd.append("fromdb", fromdb);
+        copydbCmd.append("todb", todb);
+
+        // If we don't have a username, or if we're copying locally,
+        // just run the command without authenticating
+        if ( username == "" || fromhost == "" ) {
+            return runCommand("admin", copydbCmd.done(), *info);
+        }
+
+        // Otherwise, take or guess the auth mechanism
+        string authMech;
+        if ( strcmp( mechanism.c_str(), "DEFAULT" ) != 0 ) {
+            if ( strcmp( mechanism.c_str(), "MONGODB-CR" ) != 0 &&
+                 strcmp( mechanism.c_str(), "SCRAM-SHA-1" ) != 0 ) {
+                throw UserException( 13422,
+                                     "auth mechanism must be MONGODB-CR or SCRAM-SHA-1" );
+            }
+            authMech = mechanism;
+        }
+#ifdef MONGO_SSL
+        else if ( (static_cast<DBClientBase*>(this))->getMaxWireVersion() >= 3 ) {
+            authMech = "SCRAM-SHA-1";
+        }
+#endif
+        else {
+            authMech = "MONGODB-CR";
+        }
+
+        if ( strcmp( authMech.c_str(), "MONGODB-CR" ) == 0 ) {
+            // run MONGODB-CR copydb
+            BSONObj nonceInfo;
+            BSONObjBuilder nonceCmd;
+            BSONElement e;
+            string nonce;
+
+            nonceCmd.append("copydbgetnonce", 1);
+            nonceCmd.append("fromhost", fromhost);
+
+            verify( runCommand("admin", nonceCmd.done(), nonceInfo) );
+            {
+                BSONElement e = nonceInfo.getField("nonce");
+                verify( e.type() == String );
+                nonce = e.valuestr();
+            }
+
+            copydbCmd.append("username", username);
+            copydbCmd.append("nonce", nonce);
+            copydbCmd.append("key", _authKeyCopyDB(username, password, nonce));
+            return runCommand("admin", copydbCmd.done(), *info);
+        }
+        else {
+            // run SCRAM-SHA-1 copydb, but only with SSL
```

Comment:
We'll only hit this case if the user provides "SCRAM-SHA-1" as a mechanism (so we don't have to infer it) and the driver is built without SSL.